File: lcci/0502-bianry-number-to-string/solution.cc
class Solution {
    using i64 = long long;
    static constexpr i64 D = 10000000LL;
public:
    string printBin(double num) {
        set<pair<i64, i64>> s;
        pair<i64, i64> f {static_cast<i64>(num * 2 * D), D};
        string ans = "0.";
        while (f.first != 0) {
            if (s.find(f) != s.end()) return "ERROR";
            s.insert(f);
            if (f.first >= f.second) {
                ans.push_back('1');
                f.first -= f.second;
            } else {
                ans.push_back('0');
            }
            f.first *= 2;
        }
        return ans;
    }
};

File: lcci/0502-bianry-number-to-string/solution.rs
use std::collections::HashSet;

#[derive(Hash, Eq, PartialEq, Clone)]
struct Frac {
    pub n: i64,
    pub d: i64,
}

impl Frac {
    fn new(n: f64) -> Frac {
        static d: i64 = 10000000;
        Frac {
            n: (n * d as f64) as i64,
            d,
        }
    } 
}

impl Solution {
    pub fn print_bin(num: f64) -> String {
        let mut num = Frac::new(num * 2f64);
        let mut ans = "0.".to_string();
        let mut memo = HashSet::new();
        while num.n != 0 {
            if memo.contains(&num) {
                return "ERROR".into();
            }
            memo.insert(num.clone());
            if num.n >= num.d {
                ans.push('1');
                num.n -= num.d;
            } else {
                ans.push('0');
            }
            num.n *= 2;
        }
        ans
    }
}

